### DISCUSSION TITLE
Deno Support

### DISCUSSION THREAD

**Discussion by KnorpelSenf** [vuepress/core#1045] [Ideas]
Currently, VuePress requires a Node.js setup. However neither the generated static content nor the client-side JS depends on Node in any way.

[Deno](https://deno.land)  is a new JavaScript and TypeScript runtime and it has pretty cool APIs that I'd love to use when developing plugins in order to customise my site. It would also speed up build time because the code no longer has to be compiled, since Deno can execute TypeScript.

What's your opinion on Deno support, be it through a migration/port or by creating a fork/second version of VuePress? How hard would this even be?

**Comment by Mister-Hope** [comment]
My first questions is: Are you sure all the upperchain tools support deno? E.g.: Vite.

You'd better make sure this point, before we evaluate deno support.

In my mind, deno split up its ecosystem with npm, and while vuepress are using 1000+ npm packages, we can not "Support Deno" by our own.

**Reply by KnorpelSenf** [reply]
Maybe titling this question Port to Deno was misleading, sorry. The point is that I love to use Vue, and I want to have a static site generator for technical documentation with a little bit of interactivity. VuePress has served me well for some years as it's an ideal fit here.

However, it takes several minutes to render my page and deploy it. The Deno tooling does all of this in 1-3 seconds. Projects like https://lume.land and https://fresh.deno.dev illustrate this. Now I could accept to use Vue with its 60x slowdown, or I could use different, less mature, less feature-complete, less time-tested tooling which I don't know as well—and be fast.

However, technically speaking, Vue could also play a role on Deno because it isn't in itself tied to Node, only through its tooling. That means that if I could use Vue on Deno, all my problems would be solved.

**Reply by Mister-Hope** [reply]
I understand your feelings, but I don't think that deno has a complete system about bundler, style tools and so on. There are a lot of core features which can not be ported to deno. For you you only see vue, but:

- Our build processs is built on Vite and Webpack
- Our process including sass, less and stylus support
- Our markdown parser are using markdown-it and we do support addtional plugins with markdown-it plugins
- We are using lots of useful packages like chalk, ora

There is a lot more than vue. And we cannot migrate 100+ core packages directly to deno and maintain theme by ourselves. 

So I would like to say it's too early to put a such project to Deno before deno improve strength of its ecosystem

**Reply by KnorpelSenf** [reply]
So in the bigger picture, I'm not even sure a fork would be the right thing to do because I don't know how much code would be shared. It might use a completely different stack, and it might be a new project. I'd just like it if getting fast builds doesn't force me to abandon Vue.

**Reply by KnorpelSenf** [reply]
> I don't think that deno has a complete system

That's correct, this is very much in its infant stage. That's why it seems like it's time to do something and establish Vue as the first proper solution, no? :)

**Reply by KnorpelSenf** [reply]
> we cannot migrate 100+ core packages directly to deno and maintain theme by ourselves

Ah well, fair enough. Thanks!

**Reply by KnorpelSenf** [reply]
There is an example repo for how to use Vite on Deno: https://github.com/bartlomieju/vite-deno-example

**Reply by Mister-Hope** [reply]
Hi, if you read more carefully about the demo and deno docs https://www.denojs.cn/manual/npm_nodejs/compatibility_mode, you will see that the demo is using an unstable feature called "Node.js compatibility mode" which allows node packages running on deno.

In this case, we might light vuepress run on deno, you can test it and report any uncompatable code in our repo through discussion, We might improve it if we do have a workaround. 

Also any upsteam imcompatable issue is not something we are currently insterested in.

**Reply by Mister-Hope** [reply]
Or, more clearly, if we do write something that can not run on deno, and we do have another way to implement it and let it support deno, we will make the change if you can point out the issue clearly. Feedback and contributions are welcomed, but deno support is not in our road map. We are not insterested in updating our toolchains to make sure they are "Deno compatable" and we don't want to make progress and adding test to support deno.
